frame/lottery/src/lib.rs

```diff
@@ -0,0 +1,432 @@
+		};
+		valid_calls.iter().any(|c| call_index == *c)
+	}
+}
+
+decl_storage! {
+	trait Store for Module<T: Config> as Lottery {
+		LotteryIndex: u32;
+		/// The configuration for the current lottery.
+		Lottery: Option<LotteryConfig<T::BlockNumber, BalanceOf<T>>>;
+		/// Users who have purchased a ticket. (Lottery Index, Tickets Purchased)
+		Participants: map hasher(twox_64_concat) T::AccountId => (u32, Vec<CallIndex>);
+		/// Total number of tickets sold.
+		TicketsCount: u32;
+		/// Each ticket's owner.
+		///
+		/// May have residual storage from previous lotteries. Use `TicketsCount` to see which ones
+		/// are actually valid ticket mappings.
+		Tickets: map hasher(twox_64_concat) u32 => Option<T::AccountId>;
+		/// The calls stored in this pallet to be used in an active lottery if configured
+		/// by `Config::ValidateCall`.
+		CallIndices: Vec<CallIndex>;
+	}
+}
+
+decl_event!(
+	pub enum Event<T> where
+		<T as frame_system::Config>::AccountId,
+		Balance = BalanceOf<T>,
+	{
+		/// A lottery has been started!
+		LotteryStarted,
+		/// A new set of calls have been set!
+		CallsUpdated,
+		/// A winner has been chosen!
+		Winner(AccountId, Balance),
+		/// A ticket has been bought!
+		TicketBought(AccountId, CallIndex),
+	}
+);
+
+decl_error! {
+	pub enum Error for Module<T: Config> {
+		/// An overflow has occurred.
+		Overflow,
+		/// A lottery has not been configured.
+		NotConfigured,
+		/// A lottery is already in progress.
+		InProgress,
+		/// A lottery has already ended.
+		AlreadyEnded,
+		/// The call is not valid for an open lottery.
+		InvalidCall,
+		/// You are already participating in the lottery with this call.
+		AlreadyParticipating,
+		/// Too many calls for a single lottery.
+		TooManyCalls,
+		/// Failed to encode calls
+		EncodingFailed,
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Config> for enum Call where origin: T::Origin, system = frame_system {
+		const ModuleId: ModuleId = T::ModuleId::get();
+		const MaxCalls: u32 = T::MaxCalls::get() as u32;
+
+		fn deposit_event() = default;
+
+		/// Buy a ticket to enter the lottery.
+		///
+		/// This extrinsic acts as a passthrough function for `call`. In all
+		/// situations where `call` alone would succeed, this extrinsic should
+		/// succeed.
+		///
+		/// If `call` is successful, then we will attempt to purchase a ticket,
+		/// which may fail silently. To detect success of a ticket purchase, you
+		/// should listen for the `TicketBought` event.
+		///
+		/// This extrinsic must be called by a signed origin.
+		#[weight =
+			T::WeightInfo::buy_ticket()
+				.saturating_add(call.get_dispatch_info().weight)
+		]
+		fn buy_ticket(origin, call: Box<<T as Config>::Call>) {
+			let caller = ensure_signed(origin.clone())?;
+			call.clone().dispatch(origin).map_err(|e| e.error)?;
+
+			let _ = Self::do_buy_ticket(&caller, &call);
+		}
+
+		#[weight = T::WeightInfo::set_calls(calls.len() as u32)]
+		fn set_calls(origin, calls: Vec<<T as Config>::Call>) {
```

Comment:
maybe a small doc saying this is to set calls in CallIndex storage, which might be used depending on ValidateCall configuration.